File: kotlin/lib-onebot/src/test/kotlin/uesugi/onebot/lib/connection/ConfigTest.kt
package uesugi.onebot.lib.connection

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class ConfigTest {

    @Test
    fun `default config has HTTP disabled`() {
        val config = OneBotConfig()
        assertFalse(config.httpEnable)
        assertEquals("0.0.0.0", config.httpHost)
        assertEquals(5700, config.httpPort)
    }

    @Test
    fun `default config has WS disabled`() {
        val config = OneBotConfig()
        assertFalse(config.wsEnable)
    }

    @Test
    fun `custom HTTP config`() {
        val config = OneBotConfig(
            httpHost = "127.0.0.1",
            httpPort = 8080,
        )
        assertEquals("127.0.0.1", config.httpHost)
        assertEquals(8080, config.httpPort)
    }

    @Test
    fun `WS config`() {
        val config = OneBotConfig(
            wsEnable = true,
            wsHost = "192.168.1.1",
            wsPort = 9000,
        )
        assertTrue(config.wsEnable)
        assertEquals("192.168.1.1", config.wsHost)
        assertEquals(9000, config.wsPort)
    }

    @Test
    fun `HTTP POST config`() {
        val config = OneBotConfig(
            httpPostEnable = true,
            httpPostUrl = "http://example.com/events",
            secret = "my-secret",
        )
        assertTrue(config.httpPostEnable)
        assertEquals("http://example.com/events", config.httpPostUrl)
        assertEquals("my-secret", config.secret)
    }

    @Test
    fun `reverse WS config`() {
        val config = OneBotConfig(
            wsReverseEnable = true,
            wsReverseUrl = "ws://remote:6700/ws",
            wsReverseUseUniversal = true,
        )
        assertTrue(config.wsReverseEnable)
        assertEquals("ws://remote:6700/ws", config.wsReverseUrl)
        assertTrue(config.wsReverseUseUniversal)
    }

    @Test
    fun `access token config`() {
        val config = OneBotConfig(accessToken = "bearer-token-123")
        assertEquals("bearer-token-123", config.accessToken)
    }

    @Test
    fun `timeout config`() {
        val config = OneBotConfig(timeoutMs = 15000)
        assertEquals(15000L, config.timeoutMs)
    }

    @Test
    fun `self ID config`() {
        val config = OneBotConfig(selfId = 10086)
        assertEquals(10086L, config.selfId)
    }

    @Test
    fun `app metadata config`() {
        val config = OneBotConfig(appName = "MyBot", appVersion = "2.0.0")
        assertEquals("MyBot", config.appName)
        assertEquals("2.0.0", config.appVersion)
    }
}

File: go/onebot-sdk/config_test.go
package onebot

import (
	"testing"
)

func TestDefaultConfig_HostDefaults(t *testing.T) {
	cfg := DefaultConfig()

	if cfg.HTTPHost != "0.0.0.0" {
		t.Errorf("expected HTTPHost '0.0.0.0', got %q", cfg.HTTPHost)
	}
	if cfg.WSHost != "0.0.0.0" {
		t.Errorf("expected WSHost '0.0.0.0', got %q", cfg.WSHost)
	}
}

func TestDefaultConfig_PortDefaults(t *testing.T) {
	cfg := DefaultConfig()

	if cfg.HTTPPort != 5700 {
		t.Errorf("expected HTTPPort 5700, got %d", cfg.HTTPPort)
	}
	if cfg.WSPort != 6700 {
		t.Errorf("expected WSPort 6700, got %d", cfg.WSPort)
	}
}

func TestDefaultConfig_TimeoutDefaults(t *testing.T) {
	cfg := DefaultConfig()

	if cfg.Timeout != 30000 {
		t.Errorf("expected Timeout 30000, got %d", cfg.Timeout)
	}
	if cfg.ReconnectInterval != 3000 {
		t.Errorf("expected ReconnectInterval 3000, got %d", cfg.ReconnectInterval)
	}
	if cfg.HeartbeatInterval != 15000 {
		t.Errorf("expected HeartbeatInterval 15000, got %d", cfg.HeartbeatInterval)
	}
}

func TestDefaultConfig_ZeroValueDefaults(t *testing.T) {
	cfg := DefaultConfig()

	// These should default to zero values (empty string or 0)
	if cfg.AccessToken != "" {
		t.Errorf("expected AccessToken to be empty by default, got %q", cfg.AccessToken)
	}
	if cfg.Secret != "" {
		t.Errorf("expected Secret to be empty by default, got %q", cfg.Secret)
	}
	if cfg.SelfID != 0 {
		t.Errorf("expected SelfID to be 0 by default, got %d", cfg.SelfID)
	}
	if cfg.RateLimitInterval != 0 {
		t.Errorf("expected RateLimitInterval to be 0 by default, got %d", cfg.RateLimitInterval)
	}
}

func TestConfig_FieldAssignment(t *testing.T) {
	cfg := Config{
		HTTPHost:          "127.0.0.1",
		HTTPPort:          8080,
		WSHost:            "127.0.0.1",
		WSPort:            8081,
		Timeout:           5000,
		ReconnectInterval: 1000,
		HeartbeatInterval: 5000,
		AccessToken:       "test-token",
		Secret:            "test-secret",
		SelfID:            10001,
		AppName:           "test-app",
		AppVersion:        "1.0.0",
	}

	if cfg.HTTPHost != "127.0.0.1" {
		t.Errorf("expected HTTPHost '127.0.0.1', got %q", cfg.HTTPHost)
	}
	if cfg.HTTPPort != 8080 {
		t.Errorf("expected HTTPPort 8080, got %d", cfg.HTTPPort)
	}
	if cfg.AccessToken != "test-token" {
		t.Errorf("expected AccessToken 'test-token', got %q", cfg.AccessToken)
	}
	if cfg.SelfID != 10001 {
		t.Errorf("expected SelfID 10001, got %d", cfg.SelfID)
	}
	if cfg.AppName != "test-app" {
		t.Errorf("expected AppName 'test-app', got %q", cfg.AppName)
	}
	if cfg.AppVersion != "1.0.0" {
		t.Errorf("expected AppVersion '1.0.0', got %q", cfg.AppVersion)
	}
}

func TestConfig_WebSocketReverseDefaults(t *testing.T) {
	cfg := DefaultConfig()

	if cfg.WSReverseUseUniversal != false {
		t.Errorf("expected WSReverseUseUniversal to be false by default")
	}
	if cfg.WSReverseURL != "" {
		t.Errorf("expected WSReverseURL to be empty by default, got %q", cfg.WSReverseURL)
	}
	if cfg.WSReverseAPIURL != "" {
		t.Errorf("expected WSReverseAPIURL to be empty by default, got %q", cfg.WSReverseAPIURL)
	}
	if cfg.WSReverseEventURL != "" {
		t.Errorf("expected WSReverseEventURL to be empty by default, got %q", cfg.WSReverseEventURL)
	}
}

func TestConfig_HTTPPostDefaults(t *testing.T) {
	cfg := DefaultConfig()

	if cfg.HTTPPostURL != "" {
		t.Errorf("expected HTTPPostURL to be empty by default, got %q", cfg.HTTPPostURL)
	}
	if cfg.HTTPPostHost != "" {
		t.Errorf("expected HTTPPostHost to be empty by default, got %q", cfg.HTTPPostHost)
	}
	if cfg.HTTPPostPort != 0 {
		t.Errorf("expected HTTPPostPort to be 0 by default, got %d", cfg.HTTPPostPort)
	}
}
